accepted/2020/linking-libraries.md

```diff
@@ -1,170 +1,572 @@
-
-3. In some cases the API itself is not able to be linker-friendly. In these cases we don't want to suppress the warning completely, but instead pass the warning to the caller of the API. In these cases we can use the `[RequiresUnreferencedCode]` attribute. This attribute tells the linker to not warn about this method, but instead warn about any callers of this method.
-
-4. In cases where the linker is not able to infer the dependency, or the developer wants to be explicit, the `[DynamicDependency]` attribute states an explicit dependency from one method to another. When the linker sees the attributed method needs to be kept, it will also preserve anything that method has a dynamic dependency to.
-
-A high-level goal is that the .NET Libraries are clean of linker warnings. This is one level of confidence we can have that .NET applications will be linker safe.
+The trimming tool has built-in knowledge of some of the most common problematic
+patterns and relies on annotations like `RequiresUnreferencedCodeAttribute` or
+`DynamicallyAccessedMembersAttribute` to determine the rest. If it encounters
+a pattern which it can't recognize or handle for other reasons it generates
+a trim analysis warning.
+
+Such warning tells the user that there's code in the application which may
+require additional dependency which the tool could not figure out and thus
+it's possible the app will not work correctly after trimming.
+
+See the [Linker Reflection Flow Design](https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md)
+for more information on how the process works.
+
+#### Annotating code
+
+Some library code can use reflection or other problematic patterns and thus
+running the trimming tool on it can break its functionality and it generates
+trim analysis warnings.
+
+Warnings originating in the library code itself are not actionable by the end
+user as they frequently point to internal implementation details and it's
+typically hard to determine which part of the application relies on
+the affected functionality.
+
+For this reason, affected libraries need to be annotated to help trimming tools
+to better understand the code and to generate warnings which are actionable
+in the context of an application. The goal is for the library code itself to
+not generate any warnings (as those are not actionable) and instead either
+resolve them (by providing annotations which help the trimming tool correctly
+recognize all dependencies) or to annotate public APIs such that the problem
+can be resolved in the call site by the tool automatically or the warning
+is generated pointing at the user code which uses the affected API.
+
+Depending on the specific problematic pattern different solutions can be employed.
+The list below is not complete list, it's just the most common patterns we've seen
```

Comment:
The headers of each section below describe the solution rather than the pattern/problem.